src/runtime_src/core/common/pskernel_parse.cpp

```diff
@@ -0,0 +1,230 @@
+	    } else if (dwarf_attr (type, DW_AT_bit_size, &bsize) == NULL
+		     || dwarf_formudata (&bsize, &bits) != 0)
+	      throw std::runtime_error("base type without byte or bit size");
+
+	    auto result = typeTable.find(std::make_pair(enctype,bits));
+	    ffi_type *enc_ffi_type = result->second;
+	    switch(enctype) {
+	      case DW_ATE_signed_char:
+		arg.hosttype = "int8_t";
+		break;
+	      case DW_ATE_unsigned_char:
+		arg.hosttype = "uint8_t";
+		break;
+	      case DW_ATE_signed:
+		if(bits==2)
+		  arg.hosttype = "int16_t";
+		if(bits==4)
+		  arg.hosttype = "int";
+		if(bits==8)
+		  arg.hosttype = "int64_t";
+		break;
+	      case DW_ATE_unsigned:
+		if(bits==2)
+		  arg.hosttype = "uint16_t";
+		if(bits==4)
+		  arg.hosttype = "uint32_t";
+		if(bits==8)
+		  arg.hosttype = "uint64_t";
+		break;
+	      case DW_ATE_float:
+		if(bits==4)
+		  arg.hosttype = "float";
+		if(bits==8)
+		  arg.hosttype = "double";
+		break;
+	    }
+	    arg.ffitype = *enc_ffi_type;
+	    arg.offset = offset;
+	    arg.type = xrt_core::pskernel::kernel_argument::argtype::scalar;
+	    offset = offset + bits;
+	    arg.index = index;
+	    index++;
+	  } else if (dwarf_tag(type) == DW_TAG_pointer_type) {
+	    arg.size = 16;
+	    arg.offset = offset;
+	    offset = offset + 16;  // Next offset increase by 3 words to account for 64-bit address and 64-bit size
+	    arg.index = index;
+	    arg.type = xrt_core::pskernel::kernel_argument::argtype::global;
+	    ffi_type *enc_ffi_type = &ffi_type_pointer;
+	    arg.ffitype = *enc_ffi_type;
+	    index++;
+	  }
+	  if(dwarf_aggregate_size(type, &var_size) < 0)
+	    throw std::runtime_error("ERROR: Variable Size incorrect!\n");
+
+	  return_args.emplace_back(arg);
+	  break;
+	} while (dwarf_siblingof (&child, &child) == 0);
+
+  return return_args;
+}
+
+// Function to parse object from file on disk
+std::vector<xrt_core::pskernel::kernel_argument> pskernel_parse(const char *so_file, const char *func_name) {
+
+  int fd = open(so_file, O_RDONLY);
+  Dwarf *dw = dwarf_begin(fd, DWARF_C_READ);
+  std::vector<xrt_core::pskernel::kernel_argument> args;
+
+  if (dw != NULL) {
+    Dwarf_Off offset;
+    Dwarf_Off old_offset = 0;
+    size_t hsize;
+    Dwarf_Off abbrev;
+    uint8_t addresssize;
+    uint8_t offsetsize;
+    
+    while (dwarf_nextcu (dw, old_offset, &offset, &hsize, &abbrev, &addresssize, &offsetsize) == 0) {
+      Dwarf_Die cudie_mem;
+      Dwarf_Die *cudie = dwarf_offdie (dw, old_offset + hsize, &cudie_mem);
+      Dwarf_Die child;
+      Dwarf_Die *func_die;
+      
+      if (dwarf_child (cudie, &child) == 0)
+	do {
+	  if(dwarf_tag(&child) == DW_TAG_subprogram) {
+	    const char *name = dwarf_diename (&child);
+	    if (fnmatch(func_name,name,0) == 0) {
+	      func_die = &child;
+	      args = extract_args (func_die);
+	    }
+	  }
+	} while (dwarf_siblingof (&child, &child) == 0);
+      
+      old_offset = offset;
+    }
+    
+    dwarf_end(dw);
+  }
+
+  if(!args.empty())
```

Comment:
This code should be refactor to something like this:
```
 if(args.empty())
    throw std::runtime_error("No PS kernel arguments found!");

 return args;
```

It just makes it easier to read and straight forward regarding your intent.